source/MaterialXGenShader/OCIOColorManagementSystem.h

```diff
@@ -0,0 +1,91 @@
+//
+// TM & (c) 2021 Lucasfilm Entertainment Company Ltd. and Lucasfilm Ltd.
+// All rights reserved.  See LICENSE.txt for license.
+//
+
+#ifndef MATERIALX_OCIO_COLOR_MANAGEMENT_SYSTEM_H
+#define MATERIALX_OCIO_COLOR_MANAGEMENT_SYSTEM_H
+
+/// @file
+/// Color management system which used the OCIO library
+
+#include <MaterialXFormat/File.h>
+#include <MaterialXGenShader/Nodes/SourceCodeNode.h>
+#include <MaterialXGenShader/ColorManagementSystem.h>
+
+namespace MaterialX
+{
+
+class OCIOInformation;
+
+/// A shared pointer to a OCIOColorManagementSystem
+using OCIOColorManagementSystemPtr = shared_ptr<class OCIOColorManagementSystem>;
+
+/// @class OCIOColorManagementSystem
+/// Class for a color management system which uses OCIO.
+///
+class MX_GENSHADER_API OCIOColorManagementSystem : public ColorManagementSystem
+{
+  public:
+    virtual ~OCIOColorManagementSystem();
+
+    /// Create a new OCIOColorManagementSystem
+    static OCIOColorManagementSystemPtr create(const string& target);
+
+    void loadLibrary(DocumentPtr document) override;
+
+    /// Returns whether this color management system supports a provided transform
+    bool supportsTransform(const ColorSpaceTransform& transform) const override;
+
+    /// Create a node to use to perform the given color space transformation.
+    ShaderNodePtr createNode(const ShaderGraph* parent, const ColorSpaceTransform& transform, const string& name,
+                             GenContext& context) const override;
+
+    /// Read a configuration file
+    bool readConfigFile(const FilePath& configFile);
+
+    /// Get the configuration file name
+    const FilePath& getConfigFile() const
+    {
+        return _configFile;
+    }
+
+    /// Return the OCIOColorManagementSystem name
+    const string& getName() const override
+    {
+        return OCIOColorManagementSystem::CMS_NAME;
+    }
+
+    static const string CMS_NAME;
+
+  protected:
+    /// Returns an implementation for a given transform
+    ImplementationPtr getImplementation(const ColorSpaceTransform& transform) const override;
+
+    /// Protected constructor
+    OCIOColorManagementSystem(const string& target);
+
+    /// Setup is valid
+    bool isValid() const;
+
+  private:
+    OCIOInformation*  _ocioInfo;
+    FilePath          _configFile;
+    string            _target;
+};
+
+
+/// Extending the SourceCodeNode with requirements for OCIO.
+class MX_GENSHADER_API OCIOSourceCodeNode : public SourceCodeNode
```

Comment:
From the code above, in `OCIOColorManagementSystem::getImplementation`, it seems the source code for the node is retrieved dynamically in runtime from the OCIO system. If that is the case an alternative design here would be to use a ShaderNodeImpl with dynamic code generation instead of deriving from SourceCodeNode.

I think this would simplify things. At least one advantage is that you could handle the color3->color4->color3 conversions explicitly without having to insert extra conversion nodes.

Reply:
I had originally tried put the conversion directly inside but it was much easier to just create the graphs as every shading language does the conversion differently. Also it is indeed just a single source string that is returned so is most like the source code node -- and hence the type conversions live outside. 

Perhaps moving to a general ShaderNodeImpl will come later on as it's under discussion with Patrick how to a have a more modular breakdown of the source code.